### PR TITLE
[SPARK-34595][SQL] DPP support RLIKE

### PR DESCRIPTION
**Pull request by chaojun-zhang**
### What changes were proposed in this pull request?
This pr make DPP support RLIKE expression:

```sql
SELECT date_id, product_id FROM fact_sk f
JOIN dim_store s
ON f.store_id = s.store_id WHERE s.country RLIKE  '[DE|US]'
``` 
 ### Why are the changes needed?
Improve query performance.
 
 ### Does this PR introduce _any_ user-facing change?
 No.
 
 ### How was this patch tested?
 Unit test.

### CODE REVIEW

**Comment by maropu** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 1]
```diff
@@ -18,7 +18,6 @@
 package org.apache.spark.sql
 
 import org.scalatest.GivenWhenThen
-
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:21]: please revert this.

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 2]
```diff
@@ -1403,6 +1402,52 @@ abstract class DynamicPartitionPruningSuiteBase
       )
     }
   }
+
+  test("SPARK-34436: DPP support Like/RLike expression") {
+
+
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1408]: please remove the two unnecessary blanks above.

File: sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala [hunk 3]
```diff
@@ -1403,6 +1402,52 @@ abstract class DynamicPartitionPruningSuiteBase
...
+      val df = sql(
+        """
+          |SELECT date_id, product_id FROM fact_sk f
+          |JOIN dim_store s
+          |ON f.store_id = s.store_id WHERE s.country RLIKE  '[DE|US]'
+        """.stripMargin)
+
+      checkPartitionPruningPredicate(df, false, true)
+
+
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:1436]: nit: please remove the single blank above.

**Review by c21 — COMMENTED** [review]
@chaojun-zhang - I think this PR is legit besides @maropu's comment. Wondering why close it?

**Comment by chaojun-zhang** [comment]
> @chaojun-zhang - I think this PR is legit besides @maropu's comment. Wondering why close it?

Sorry,  there is another  PR  I created below that is the same as this one,  actually, I should close that one, so just ignore this   
one.

https://github.com/apache/spark/pull/31722

**Pull request closed without merging**